### ISSUE TITLE
error in oi_im_flush_kerb() ?

### ISSUE THREAD

**Issue by GretaTimaite** [udsleeds/openinfra#141]
Happy to see the function being split @hulsiejames , however I think there's a mistake in the im_oi_flush_kerb function or its name?

The function name has "im" in it, so I assume it is based on the inclusive mobility guide, e.g., the function I've wrote some time ago.

Currently it [looks like this](https://github.com/udsleeds/openinfra/commit/104b5087f7bec538c167223d98f5bdce000cffde#diff-22ff88cefa3672f9dcbbe6c7fb864cf212a680ffff55714682bbac91b895d6d1):

```{r}
oi_im_flush_kerb = function(osm_sf){

  osm_sf_im = osm_sf |>
    # Assesses the presence of a kerb, and what type of kerb it is.
    dplyr::mutate(openinfra_im_kerb = dplyr::case_when(
      stringr::str_detect(kerb, "lower") ~ "lowered kerb",
      kerb == "flush" ~ "flush kerb",
      kerb %in% c("no", "none", "flat") ~ "no kerb", 
      kerb == "raised" ~ "raised kerb",
      kerb == "rolled" ~ "rolled kerb",
      kerb %in% c("yes", "regular", "normal") ~ "normal kerb",
      (TRUE & !is.na(kerb)) ~ "unknown",
      is.na(kerb) ~ "missing data"
    )) |>
    dplyr::mutate(openinfra_im_accessible_kerb = dplyr::case_when(
      # Based on assessments above, assesses whether the kerb is accessible
      # for wheelchair users. Definitions decided based on kerb wiki below.
      # See: https://wiki.openstreetmap.org/wiki/Key:kerb
      openinfra_im_kerb %in% c("lowered kerb", "flush kerb",
                               "no kerb") ~ "wheelchair - yes",
      openinfra_im_kerb %in% c("raised kerb", "rolled kerb", "normal kerb"
                               ) ~ "wheelchair - no",
      openinfra_im_kerb %in% c("unknown", "missing data") ~ "lacking data"
    ))

  return(osm_sf_im)
  }
```

Based on the function it looks to me that it is a "regular" recategorisation of kerbs based on the OSM wiki rather than Inclusive Mobility. Function description also says: ["Function to assess whether a kerb (where tagged) is flush or not."](https://udsleeds.github.io/openinfra/reference/oi_im_flush_kerb.html).  

Basically what I want to point out is that it's misleading to have 'im" when it's not what the function does? The key difference is that the OSM understanding of "wheelchair-accessible" kerb diverges from IM understanding. I think it is clearly noted in the [vignette](https://udsleeds.github.io/openinfra/articles/im_get.html#justification): "while there is an implication in OSM tagging scheme that kerbs of ~ 3cm height are wheelchair-friendly, it is not a case in the IM guide. Therefore, only kerbs having values of ‘flush’ or ‘no’ have been categorized as “flush”." 

In other words, _lowered kerb_ would _not_ be accessible based on the IM guide.

Also out of curiosity, why do you create a new `openinfra_im_kerb` object within your function? Is it a normal practice? Because it looks like an unnecessary step to me in this context.

**Comment by GretaTimaite** [comment]
Also, I haven't examined the rest of the functions, but what I've noticed by skimming through the code is the use of both base R and tidyverse pipes. Maybe picking one would be better? Just for code coherence

**Comment by GretaTimaite** [comment]
To wrap up (apologies for multiple entries -- my brain is scattered today), based on IM guide, no kerbs are accessible unless it has "flush" or "no"/"none/"flat" values.  I think having simple Boolean yes/no is a way to go. Missing data is always a subject to uncertainty (does it exist or not?), so maybe does not need to be mentioned explicitly. However, it's not something I'm bothered about.

**Comment by GretaTimaite** [comment]
Do let me know if I just misunderstood (didn't see?) something in this function @hulsiejames. 👍🏻

**Comment by hulsiejames** [comment]
Hi Greta, 

I had been using the im_ noation in the new split functions to show that they were once part of the single IM function that we have split up #129. 

My thinking behind this was, I didn't want someone to see previous outptus/reports where we have made reference to an IM function, and now (when I will remove it from {openinfra} functions) readers may be confused to not find an IM function. 
___________________________________________________________________
>The key difference is that the OSM understanding of "wheelchair-accessible" kerb diverges from IM understanding. I think it is clearly noted in the [vignette](https://udsleeds.github.io/openinfra/articles/im_get.html#justification): "while there is an implication in OSM tagging scheme that kerbs of ~ 3cm height are wheelchair-friendly, it is not a case in the IM guide. Therefore, only kerbs having values of ‘flush’ or ‘no’ have been categorized as “flush”."

I see, thank you for the clarifications. To be honest I was splitting the function up looking at the source code and so forgot to reference the justification vignette! (I have some more thoughts on this I wil discuss at the bottom of this comment!)
__________________________________________________________
>Also out of curiosity, why do you create a new openinfra_im_kerb object within your function? Is it a normal practice? Because it looks like an unnecessary step to me in this context.

Assuming the comment is making reference to this `dplyr::mutate(openinfra_im_kerb = dplyr::case_when(`, this is specifying the column name that will appear within the `osm_sf` once the function has been applied, the column values are then decided/created by the `case_when`. So the column name `operninfra_im_kerb` will be added by the `oi_im_flush_kerb` function.

`oi_im_flush_kerb = function(osm_sf){func code here}` specifies the name of the package function to be called, i.e. `openifnra::oi_im_flush_kerb()` will call the function code contained within {}.

With `openinfra_im_kerb %in% c("lowered kerb", "flush kerb", "no kerb") ~ "wheelchair - yes" ` I am asking to check column values of `openinfra_im_kerb`, established by the previous `case_when`, to make assessments on whether the infrastructure would be accessible to wheelchairs. 
__________________________________________________________________

>EDIT: I assume this is a result of you copying old code when base R pipe was not out yet and adding new code with base R pipe.

This is indeed the case! Thank you for the spot, I will go and change them to new pipe (I prefer) for consistency
_________________________________________________________________
>Missing data is always a subject to uncertainty (does it exist or not?), so maybe does not need to be mentioned explicitly.

I have added this classification to be able to distinguish why a feature has not beed assessed. This is either due to:

**A)** the `kerb` field is NA - this does not imply the presence of lack of kerb, it lets the user know OSM has no data on this. As such, one may wish to go and collect data from such an area or organise a mapathon. It allows users to assess lacking coverage of OSM data.

**B)** the `kerb` field is not NA, but is not a value that is recategorisable to a distinctive kerb value (such as flush, raised etc). This lets a user know that data has been entered into OSM, but perhaps it is an incorrect value and the user may go and assess this. Alteratively, it is a correct value and we have missed this in our code, in which case I'd hope the user would create an issue on the repo and we can add this.

Technically, I have sort of added this as a debugging step if you like. It makes no assumption on the presence of lack of kerb infrastructure. I will go and clarify this in the docs.
_________________________________________________________________
>(I have some more thoughts on this I wil discuss at the bottom of this comment!)

Is this something that requires it's own function (specifically the identifiation of IM friendly kerbs)? 
I say this as it seems one could find the IM kerbs by applying the following line of code: 
```
osm_im_kerbs = osm_data %>% dplyr::filter(kerb %in% c("flush", "no", "none", "flat"))
```

If this is a reasonable idea, how about we keep the functionality of this function the same but rename it from `oi_im_flush_kerb` to `oi_assess_kerbs`. We could then say...

_"To identify kerbs that are accessible by IM Guide definitions, apply the `oi_assess_kerbs` function, and query the output column titled `openinfra_kerbs` (`openinfra_im_kerb` -> `openinfra_kerbs`) for the values "flush kerb" and "no kerb"."_

That is: 
```
osm_im_kerbs = osm_data %>% oi_assess_kerbs() %>% dplyr::filter(openinfra_kerbs %in% c("flush kerb", "no kerb"))
```

**Comment by GretaTimaite** [comment]
Apologies for not returning to this earlier. 

----
> Assuming the comment is making reference to this dplyr::mutate(openinfra_im_kerb = dplyr::case_when(

Yes, I was inquiring about the purpose of `openinfra_im_kerb` column. 
I was looking at the function from the IM guide perspective, so I wasn't sure why one needs an extra column (makes function more complex, imo) as these could be combined. However,  if you want this function to be a general recategorisation of OSM kerb tag then it's fine. 

----
> I have added this classification to be able to distinguish why a feature has not beed assessed.

What I meant by this is that I think that returning NA value seems enough to encompass both instances. I think this is what implies missing/unavailability of (enough) data to categorise them. However, this is a minor detail and, I'd say, a matter of consensus of indicating things.
---

> Is this something that requires it's own function (specifically the identifiation of IM friendly kerbs)?

Yes OR the `im` should be removed from `oi_im_flush_kerb`. You essentially change (not just tweak it) what the function does: the way you define the function is not consistent with the idea of inclusive mobility function or what inclusive mobility requirements for kerbs are. IMO, this is fundamentally misleading. 

> I say this as it seems one could find the IM kerbs by applying the following line of code

yes, it seems easy BECAUSE you know the IM guidelines. To write this (simple) function you need to consult both OSM and IM. In my opinion it requires more knowledge and effort than only OSM (which has a well documented wiki on kerbs).

In other words, just because a code is simple/easy to read, it doesn't mean it's easy to write it. 

I don't mind further refining/homogenizing OSM kerb values, but it must be separated from the IM as they are not compatible because of different understanding of what is wheelchair accessible. This is the main point of the whole issue I raised.

---
> I'm not sure of the best appraoch here, given the aim was to split the origional IM function into individual functions, but we also want to maintain a good balance of useful funtctions rather than functions for the sake of it?

I agree with the idea but I'd rather have a few extra ones to mak everything crystal clear.

Also I'm not completely sure of the context in which brought up tactile paving? Does it need its own function or not?

**Comment by hulsiejames** [comment]
Hi Greta, sorry for the delayed response!

>Yes, I was inquiring about the purpose of openinfra_im_kerb column **&&&**  However, if you want this function to be a general recategorisation of OSM kerb tag then it's fine

In a sense this is correct, I had since renamed `openinfra_im_kerb` to `openinfra_kerb` as this is not an indication of kerbs that are IM accessible, more this is a general recategorisation of OSM kerb values.

However, I am going to remove this. As kerb=* values (at least for england) are quite descriptive and self-explanatory, thus `openinfra_kerb` is no longer required. 

With this, I will re-create the `openinfra_im_flush_kerb` column within the function as it used to be, with the addition that the kerb [must have tactile paving](https://udsleeds.github.io/openinfra/articles/im_get.html#kerb).
That is, for a kerb to be accessible by IM guidance, it must be: `kerb="no"|"flush" & tactile_paving="yes"`. 

In England there are 211 ways that meet the requirement for `kerb="no"|"flush"`,
moreover, there are 159 ways that meet the requirements for `kerb="no"|"flush" & tactile_paving="yes"` - an IM compliant flush kerb.

A pleasnt surprise is that Leeds has the best coverage in England! 
![Screenshot from 2022-10-08 12-58-31](https://user-images.githubusercontent.com/58815827/194706339-4927691c-deaf-45fa-9e74-11b83bf76b2a.png)

_____________________________________________________

>What I meant by this is that I think that returning NA value seems enough to encompass both instances.

To be fair I do agree with this. 

My only reason for including it is that someone using the datapacks may not be fully aware of the implications of NA data in this context and assume it to imply "no" rather than "maybe - but has not been entered to OSM so cannot assess", as you say is a matter of consensus! 

_____________________________________________________

I have changed this back now as in agreement, though: 

>To write this (simple) function you need to consult both OSM and IM.   

In agreement with this.  

I think what I was trying to say is what you have put here, it requires consultation of the OSM wiki and IM guide, which could be described textually rather than code. But this does require one to know where to look (specifically OSM wiki and IM guidance) and so is valuable, I was in the mindset of trying to cut down a potentially unnecessary number of functions. 

_____________________________________________________

> Also I'm not completely sure of the context in which brought up tactile paving? Does it need its own function or not?

I am 50/50 on this. Origionally I was going to include this within the `oi_im_flush_kerb` function as you need to know the presence of tactile paving to have a compliant flush kerb. However, users may not realsie that the assessment of tactile paving is made by the `oi_im_flush_kerb` function.

So, I think for now I will leave these as two seperate functions and just call `oi_im_tactile_pacing` within the `oi_im_flush_kerb` function.
